### PR TITLE
[loss] Redesign loss abstractions 

### PR DESCRIPTION
**Pull request by acisseJZhong**
## Motivation

The current loss hierarchy lacks a clear separation between interfaces, input
contracts, loss composition, and execution:

- `BaseLoss`, `CrossEntropyLoss`, and `MSELoss` define a tensor-only input
  contract. However, `MTPLoss` inherits from `CrossEntropyLoss` while requiring
  aligned tuples at runtime, and `ChunkedLossWrapper` inherits from `BaseLoss`
  while accepting either tensors or tuples depending on its inner loss. The
  hierarchy therefore does not accurately describe the inputs accepted by each
  loss.
- `BaseLoss` is both the trainer-facing interface and a default single-output
  loss implementation.
- `CrossEntropyLoss` largely duplicates the `BaseLoss` path to pass
  `global_vocab_size` to the cross-entropy kernel.
- `MTPLoss` combines multiple cross-entropy terms, but inherits from
  `CrossEntropyLoss` as if it were a specialized cross-entropy implementation.
- `ChunkedLossWrapper` changes how another loss is executed, but inherits from
  `BaseLoss` as if it defined a new objective.

This makes input contracts unclear and incorrectly represents loss composition
and execution wrapping as implementation inheritance.

### 1. `Loss`: a typed callable interface

`Loss` is a protocol with one type parameter describing the shared input
structure of `pred` and `labels`:

```python
LossInputT = TypeVar("LossInputT", contravariant=True)

class Loss(Protocol[LossInputT]):
    def __call__(
        self,
        pred: LossInputT,
        labels: LossInputT,
        global_valid_tokens: torch.Tensor | None = None,
        **loss_inputs: Any,
    ) -> LossResult:
        ...
```

`CrossEntropyLoss` and `MSELoss` implement `Loss[torch.Tensor]`, while
`MTPLoss` implements `Loss[tuple[torch.Tensor, ...]]`. This makes their runtime
input contracts explicit without sharing an implementation base class.

### 2. `LossConfig`: configuration only

`LossConfig` remains a common, non-generic `Configurable.Config` base used for
trainer configuration and component construction. It does not attempt to carry
callable type information through the dynamic configuration system.

### 3. `MTPLoss`: composition instead of inheritance

`MTPLoss` is a standalone configurable loss that contains a
`CrossEntropyLoss`. It applies cross-entropy to the main and MTP-depth outputs,
combines the weighted terms, and normalizes the result once.

### 4. `ChunkedLossWrapper`: execution instead of objective semantics

`ChunkedLossWrapper` accepts the two structures it actually supports:

```python
LossInput = torch.Tensor | tuple[torch.Tensor, ...]
```

It validates at runtime that predictions and labels have matching structures,
then applies the configured loss to each chunk. It owns chunking, `lm_head`
execution, FSDP state management, backward, and gradient reconnection; it does
not define or interpret the wrapped objective.

The resulting relationships are composition rather than implementation
inheritance:

```text
chunked cross-entropy: ChunkedLossWrapper -> CrossEntropyLoss
chunked MTP:           ChunkedLossWrapper -> MTPLoss -> CrossEntropyLoss
```

Policy losses such as DAPO and GRPO retain their specialized keyword inputs and
are configured through the common `LossConfig` base.

## Detailed changes

- Replace `BaseLoss` with the `Loss[T]` protocol and non-generic `LossConfig`.
- Make cross-entropy and MSE implement the tensor input contract directly.
- Make `MTPLoss` contain `CrossEntropyLoss` instead of inheriting from it.
- Make `ChunkedLossWrapper` explicitly support tensor and tuple inputs while
  preserving runtime structure validation.
- Update trainer and experiment call sites to use the new loss contracts.

## Validation

Both numerical comparisons used baseline commit `310e2a66e`, the same seed
checkpoint, `--debug.seed=42`, `--debug.deterministic`, disabled CUDA graphs,
and full-precision TensorBoard values for 10 training steps.

| Configuration | GPUs | Maximum loss difference | Maximum grad-norm difference |
| --- | ---: | ---: | ---: |
| MTP + chunked loss | 1 | 0 | 0 |
| MTP + chunked loss, TP=2, CP=2, SP | 4 | 0 | 0 |

Additional validation:

- Loss, CLI-config, graph-trainer, config-manager, and CP-validation tests:
  84 passed, 7 subtests passed.
- Real-process-group `deepseek_v3_mtp_tp+cp`: 10/10 steps passed on 4 GPUs.
- Compiled MTP + chunked loss CPU smoke test using the eager compile backend
  completed forward and backward successfully.

### CODE REVIEW

File: torchtitan/experiments/rl/losses/grpo.py [hunk 1]
```diff
@@ -22,7 +22,7 @@ class GRPOLoss(DAPOLoss):
     """
 
     @dataclass(kw_only=True, slots=True)
-    class Config(BaseLoss.Config):
+    class Config(LossConfig):
```
Comment (tianyu-l) [inline torchtitan/experiments/rl/losses/grpo.py:25]: inherit `DAPOLoss.Config`

File: torchtitan/experiments/rl/losses/dapo.py [hunk 2]
```diff
@@ -12,15 +12,15 @@
...
+from torchtitan.components.loss import compute_logprobs, LossConfig
+from torchtitan.config import CompileConfig, Configurable
 
 # Clamp |log(pi_theta/pi_old)| before exp() so a large generator/trainer
 # logprob mismatch cannot overflow exp() to inf/NaN.
 _MAX_LOG_RATIO = 10.0
 
 
-class DAPOLoss(BaseLoss):
+class DAPOLoss(Configurable):
```
Comment (tianyu-l) [inline torchtitan/experiments/rl/losses/dapo.py:23]: why not inheriting Loss?

File: torchtitan/components/loss.py [hunk 3]
```diff
@@ -27,6 +27,60 @@
 IGNORE_INDEX = -100
 
 LossFunction: TypeAlias = Callable[..., torch.Tensor]
+LossResult: TypeAlias = tuple[torch.Tensor, dict[str, torch.Tensor]]
+
+LossInput: TypeAlias = torch.Tensor | tuple[torch.Tensor, ...]
+LossInputT = TypeVar("LossInputT", contravariant=True)
+
+
+class Loss(Protocol[LossInputT]):
```
Comment (tianyu-l) [inline torchtitan/components/loss.py:36]: why don't we inherit https://github.com/pytorch/torchtitan/blob/main/torchtitan/config/function.py#L16 and make `LossConfig` -> `Loss.Config`?

File: torchtitan/models/deepseek_v3/mtp.py [hunk 4]
```diff
@@ -449,21 +455,29 @@ def apply_fsdp_to_mtp_decoder(
             del model.layers[key]
 
 
-class MTPLoss(CrossEntropyLoss):
+class MTPLoss(
```
Comment (tianyu-l) [inline torchtitan/models/deepseek_v3/mtp.py:458]: If hardcoding cross-entropy, should be called `MTPCrossEntropyLoss`; o/w make it generic by owning a tensor loss config?

File: torchtitan/components/loss.py [hunk 5]
```diff
@@ -708,6 +665,60 @@ def _gradient_backprop(
         )
 
 
+class GradAccumulator:
```
Comment (tianyu-l) [inline torchtitan/components/loss.py:668]: Is this only location change?

File: torchtitan/components/loss.py [hunk 6]
```diff
@@ -473,17 +427,20 @@ def __init__(
...
+        self.loss_fn: Callable[..., LossResult] = config.loss_fn.build(
+            compile_config=compile_config
+        )
         self.lm_head: nn.Module | None = None
 
     def set_lm_head(self, lm_head: nn.Module) -> None:
         """Set the lm_head module. Must be called before the first __call__."""
         self.lm_head = lm_head
 
+    @override
```
Comment (tianyu-l) [inline torchtitan/components/loss.py:439]: why this wrapper?

File: torchtitan/components/loss.py [hunk 7]
```diff
@@ -207,109 +261,60 @@ def backward(  # pyrefly: ignore[bad-override]
...
-
-    def _maybe_compile(self, compile_config: CompileConfig | None) -> None:
-        if (
-            compile_config is not None
-            and compile_config.enable
-            and "loss" in compile_config.components
-        ):
-            logger.info("Compiling the loss function with torch.compile")
-            self.fn = torch.compile(self.fn, backend=compile_config.backend)
+        self.fn = _maybe_compile_loss_fn(cross_entropy_loss, compile_config)
```
Comment (tianyu-l) [inline torchtitan/components/loss.py:273]: why moving this out to be a function?

File: torchtitan/components/loss.py [hunk 8]
```diff
@@ -207,109 +261,60 @@ def backward(  # pyrefly: ignore[bad-override]
...
-            if global_valid_tokens is not None:
-                spmd.assert_type(
-                    global_valid_tokens,
-                    {"dp": spmd.R, "cp": spmd.R, "tp": spmd.I},
-                )
-        if global_valid_tokens is not None:
-            loss = loss / global_valid_tokens
-        return loss, {}
+        loss = self.fn(pred, labels, global_vocab_size=self.global_vocab_size)
+        return _normalize_loss(loss, global_valid_tokens), {}
```
Comment (tianyu-l) [inline torchtitan/components/loss.py:286]: similar

**Review by tianyu-l — COMMENTED** [review]
nice direction! Left some comments
